In [46]:
import numpy as np
import pandas as pd

In [50]:
data = pd.read_csv('/content/v4.csv')

In [48]:
data.shape

(1000, 19)

In [51]:
data.head()

Country      Status  Life_expectancy  \
0  Saint Vincent and the Grenadines  Developing             73.2   
1                             Chile  Developing             77.3   
2                             Egypt  Developing             69.0   
3                 Republic of Korea  Developing             81.2   
4                        Costa Rica  Developing             78.1   

   Adult_Mortality  infant_deaths  Alcohol  percentage_expenditure  \
0            156.0              0      NaN                0.000000   
1             13.0              2     6.24               74.148429   
2            176.0             57     0.16                0.000000   
3             69.0              1     9.56                0.000000   
4             16.0              1     3.58             2377.010299   

   Hepatitis_B  Measles   BMI  under-five_deaths  Polio  Total_expenditure  \
0         98.0        0  54.1                  0   97.0                NaN   
1          NaN        0  54.0                  3   91.0               6.40   
2         97.0       80  52.9                 70   97.0               5.21   
3         99.0        2   3.4                  2   99.0               7.10   
4         89.0        0  55.5                  1   93.0               9.66   

   Diphtheria  HIV_AIDS          GDP  Population  \
0        98.0       0.1          NaN         NaN   
1        91.0       0.1   511.368479  15262754.0   
2        97.0       0.1          NaN         NaN   
3        99.0       0.1          NaN         NaN   
4        88.0       0.1  8199.414621    454528.0   

   Income_composition_of_resources  Schooling  
0                            0.720       13.3  
1                            0.755       13.5  
2                            0.623       11.2  
3                              NaN        NaN  
4                            0.749       13.3

In [52]:
data.isna().sum()

Country                              0
Status                               0
Life_expectancy                      5
Adult_Mortality                      5
infant_deaths                        0
Alcohol                             50
percentage_expenditure               0
Hepatitis_B                        172
Measles                              0
BMI                                  9
under-five_deaths                    0
Polio                                5
Total_expenditure                   65
Diphtheria                           5
HIV_AIDS                             0
GDP                                145
Population                         213
Income_composition_of_resources     54
Schooling                           51
dtype: int64

In [53]:
data['Life_expectancy'].isnull().sum()

5

Q3. Are there any missing values in the column 'Life_expectancy? If there are, remove the corresponding rows from the data. Note that this modified data will be used for the subsequent questions. 


In [54]:
df = data.dropna(subset = ['Life_expectancy'])

In [55]:
df.shape

(995, 19)

Q4: [marks: 1] How many categorical features are there in the dataset? 


In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 995 entries, 0 to 999
Data columns (total 19 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          995 non-null    object 
 1   Status                           995 non-null    object 
 2   Life_expectancy                  995 non-null    float64
 3   Adult_Mortality                  995 non-null    float64
 4   infant_deaths                    995 non-null    int64  
 5   Alcohol                          945 non-null    float64
 6   percentage_expenditure           995 non-null    float64
 7   Hepatitis_B                      823 non-null    float64
 8   Measles                          995 non-null    int64  
 9   BMI                              987 non-null    float64
 10  under-five_deaths                995 non-null    int64  
 11  Polio                            990 non-null    float64
 12  Total_expenditure     

ans. 2

Q5: [marks: 1] What is the average BMI in the dataset? 


In [57]:
df['BMI'].mean()

38.063931104356634

Q6 [marks: 2] How many missing values are there in the columns 'Hepatitis_B' and 'Population? [MCQ] 

In [58]:
df.isnull().sum()

Country                              0
Status                               0
Life_expectancy                      0
Adult_Mortality                      0
infant_deaths                        0
Alcohol                             50
percentage_expenditure               0
Hepatitis_B                        172
Measles                              0
BMI                                  8
under-five_deaths                    0
Polio                                5
Total_expenditure                   65
Diphtheria                           5
HIV_AIDS                             0
GDP                                143
Population                         209
Income_composition_of_resources     50
Schooling                           50
dtype: int64

ans 172, 209

Q7: [marks: 1] How many unique countries are there in the dataset? 


In [59]:
len(df['Country'].unique())

183

In [60]:
df.head()

Country      Status  Life_expectancy  \
0  Saint Vincent and the Grenadines  Developing             73.2   
1                             Chile  Developing             77.3   
2                             Egypt  Developing             69.0   
3                 Republic of Korea  Developing             81.2   
4                        Costa Rica  Developing             78.1   

   Adult_Mortality  infant_deaths  Alcohol  percentage_expenditure  \
0            156.0              0      NaN                0.000000   
1             13.0              2     6.24               74.148429   
2            176.0             57     0.16                0.000000   
3             69.0              1     9.56                0.000000   
4             16.0              1     3.58             2377.010299   

   Hepatitis_B  Measles   BMI  under-five_deaths  Polio  Total_expenditure  \
0         98.0        0  54.1                  0   97.0                NaN   
1          NaN        0  54.0                  3   91.0               6.40   
2         97.0       80  52.9                 70   97.0               5.21   
3         99.0        2   3.4                  2   99.0               7.10   
4         89.0        0  55.5                  1   93.0               9.66   

   Diphtheria  HIV_AIDS          GDP  Population  \
0        98.0       0.1          NaN         NaN   
1        91.0       0.1   511.368479  15262754.0   
2        97.0       0.1          NaN         NaN   
3        99.0       0.1          NaN         NaN   
4        88.0       0.1  8199.414621    454528.0   

   Income_composition_of_resources  Schooling  
0                            0.720       13.3  
1                            0.755       13.5  
2                            0.623       11.2  
3                              NaN        NaN  
4                            0.749       13.3

In [61]:
target = df.loc[:,'Life_expectancy']
features = df.drop(['Life_expectancy'],axis = 1)

In [62]:
target.head()

0    73.2
1    77.3
2    69.0
3    81.2
4    78.1
Name: Life_expectancy, dtype: float64

In [65]:
target.mean()

69.29557788944724

In [66]:
from sklearn.model_selection import train_test_split
X_train, X_test,y_train, y_test = train_test_split(features, target, random_state = 42, test_size = 0.2)

In [67]:
from sklearn.preprocessing import PolynomialFeatures

poly  = PolynomialFeatures(degree = 2)
polydata = poly.fit_transform(X_train[X_train.columns[2:4]])

polydata.shape

(796, 6)

Prepare a pipeline numeric_transformer containing SimpleImputer (strategy="mean") and StandardScaler() (in this sequence). Preprocess the numeric_features of the given data ('Adult_Mortality', 'infant_deaths', 'Alcohol', 'percentage_expenditure', 'Hepatitis_B', 'Measles', 'BMI', 'under-five_deaths', 'Polio', 'Total_expenditure', 'Diphtheria', 'HIV AIDS', 'GDP', 'Population', 'Income_composition_of_resources', 'Schooling') using this pipeline. Apply 


In [68]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

numeric_features = ['Adult_Mortality', 'infant_deaths', 'Alcohol', 'percentage_expenditure', 
                    'Hepatitis_B', 'Measles', 'BMI', 'under-five_deaths', 'Polio',
                    'Total_expenditure', 'Diphtheria', 'HIV_AIDS', 'GDP', 'Population',
                    'Income_composition_of_resources', 'Schooling']

categorical_features = ['Country', 'Status']


numeric_transformer = Pipeline([('imputer', SimpleImputer(missing_values = np.nan, strategy = 'mean')),
                                ('scaler', StandardScaler())])

categorical_transformer = OneHotEncoder(handle_unknown  = 'ignore')

transformer = ColumnTransformer(transformers = [('num', numeric_transformer, numeric_features),
                                                ('cat', categorical_transformer, categorical_features)],
                                remainder = 'passthrough')
                                

transformer.fit(X_train)

X_train = transformer.transform(X_train)
X_test = transformer.transform(X_test)

In [69]:
X_train.shape

(796, 200)

In [70]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import LinearRegression

estimator = LinearRegression()
selector = RFE(estimator,
               n_features_to_select = 15
               )

selector.fit(X_train, y_train)


RFE(estimator=LinearRegression(), n_features_to_select=15)

In [71]:
selector.ranking_

array([172,   1, 161, 174, 184, 186, 185,   1, 164, 178, 173, 129, 171,
       183, 168,  47,  18,  60, 106,   1,  45, 125,  80, 154, 141, 118,
        79,  87,  78, 116, 157, 143, 159,  16, 163,  44,  62,   1, 120,
        69,  71,  58,   9,  86, 153,   7,  28,   1,   4,  57, 109,  94,
        12,   3,  43, 127,  65,  40, 169,   1, 180,  25, 177, 167,  97,
        83,  93,  92,   2, 112, 155, 140, 134,  50,  52,  21,  17,  95,
       158,  23,  49,  77,  48,  10,   1, 136,  19,  63,  76, 170,   1,
       144, 108,  59, 166,  53,  30,  74,  31, 113,  42,  20,  37, 101,
       137, 123,  67,  91,   1,   1, 179,  46,  39,  34,   1,  85,  54,
        15,  38, 122, 119,  61, 175,  33, 103,  73,   6, 146,  13, 152,
       162, 131,  75, 147, 160, 165,  66, 151,  51,  14, 104, 148,  81,
       145,  56,  26, 138,  70, 182,  27,  90,  99,  84, 135,  96, 124,
       102, 115,   1,  36, 117, 132, 130,   8,  11,  68,  32,  98, 150,
       121,   1,  29,  35,  88, 176, 100,  82, 139,   5, 128, 13

In [72]:
selector.support_

array([False,  True, False, False, False, False, False,  True, False,
       False, False, False, False, False, False, False, False, False,
       False,  True, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False,  True, False, False, False, False, False, False, False,
       False, False,  True, False, False, False, False, False, False,
       False, False, False, False, False,  True, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False,  True, False, False, False, False, False,
        True, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
        True,  True, False, False, False, False,  True, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

Fit a LinearRegression model that uses the normal equation to learn the weights on the train data (x_train and y_train). Enter the value of score obtained using training data upto four decimal places. 


In [35]:
?LinearRegression

In [73]:
regressor = LinearRegression()
regressor.fit(X_train, y_train)

LinearRegression()

In [74]:
regressor.score(X_train,y_train)

0.9668633016298744

In [75]:
from sklearn.model_selection import cross_val_score

cross_val = cross_val_score(LinearRegression(), X_train, y_train, cv = 15 )
cross_val.max()

0.9710757848173139

Q19. Fit a Stochastic Gradient Descent regressor model on the training data (X_train and y_train)\
Set the following parameters:\
(i) penalty = II\
(ii) alpha = 0.001\
(iii) learning_rate = 'constant'\
(iv) initial learning rate = 0.001\
(v) random state = 42\
Other parameters are initialized with default values.\
Compute the for training data. Using the trained model, make predictions on
test data and then compute the for the test data also. Find the absolute
difference between the training and testing errors computed and enter the value upto two decimal
places, i.e. 1M SE
- MSEtest
train


In [78]:
from sklearn.linear_model import SGDRegressor
sgdregressor = SGDRegressor(penalty = 'l1', alpha = 0.001, learning_rate = 'constant',
                            eta0 = 0.001, random_state = 42)

sgdregressor.fit(X_train, y_train)

SGDRegressor(alpha=0.001, eta0=0.001, learning_rate='constant', penalty='l1',
             random_state=42)

In [81]:
from sklearn.metrics import mean_squared_error
y_pred = sgdregressor.predict(X_test)
y_predt = sgdregressor.predict(X_train)
mse = mean_squared_error(y_test, y_pred)
mset = mean_squared_error(y_train, y_predt)
ans = np.abs(mse - mset)
ans

2.1031096193385794